src/NUnitFramework/framework/TestParameters.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace NUnit.Framework
+{
+    /// <summary>
+    /// TestParameters class holds any named parameters supplied to the test run
+    /// </summary>
+    public class TestParameters
+    {
+        private readonly Dictionary<string, string> _parameters = new Dictionary<string, string>();
+
+        /// <summary>
+        /// Gets the number of test parameters
+        /// </summary>
+        public int Count
+        {
+            get { return _parameters.Count; }
+        }
+
+        /// <summary>
+        /// Gets a collection of the test parameter names
+        /// </summary>
+        public ICollection<string> Names
+        {
+            get { return _parameters.Keys; }
+        }
+
+        /// <summary>
+        /// Gets a flag indicating whether a parameter with the specified name exists.N
+        /// </summary>
+        /// <param name="name">Name of the parameter</param>
+        /// <returns>True if it exists, otherwise false</returns>
+        public bool Exists(string name)
+        {
+            return _parameters.ContainsKey(name);
+        }
+
+        /// <summary>
+        /// Indexer provides access to the internal dictionary
+        /// </summary>
+        /// <param name="name">Name of the parameter</param>
+        /// <returns>Value of the parameter or null if not present</returns>
+        public string this[string name]
+        {
+            get { return Get(name); }
+        }
+
+        /// <summary>
+        /// Get method is a simple alternative to the indexer
+        /// </summary>
+        /// <param name="name">Name of the paramter</param>
+        /// <returns>Value of the parameter or null if not present</returns>
+        public string Get(string name)
+        {
+            return Exists(name) ? _parameters[name] : null;
+        }
+
```

Comment:
Using exists and then getting the value requires two completely independent lookups in the dictionary.  Trygetvalue only does a single lookup.  I never use the pattern you used.  As far as I know every platform has trygetvalue for the generic dictionary class.

---Sent from Boxer | http://getboxer.com

On June 17, 2016 at 9:08:04 AM GMT+8, CharliePoole <notifications@github.com> wrote:In src/NUnitFramework/framework/TestParameters.cs:  > + /// <returns>Value of the parameter or null if not present</returns> > + public string this[string name] > + { > + get { return Get(name); } > + } > + > + /// <summary> > + /// Get method is a simple alternative to the indexer > + /// </summary> > + /// <param name="name">Name of the paramter</param> > + /// <returns>Value of the parameter or null if not present</returns> > + public string Get(string name) > + { > + return Exists(name) ? _parameters[name] : null; > + } > +   I used Exists for clarity, but it's basically as if I wrote...  if (_parameters.ContainsKey(name) return _parameters[name];   This is a pattern we use all over NUnit. If memory serves, it's because TryGetValue doesn't (or maybe didn't) exist on some platforms. As far as I have always understood, it's a little bit inefficient, but not enough to make me want to prematurely optimize it, especially at the expense of clarity. The various TryXxx functions always seem very hard to read.   Do you have an estimate of how inefficient it is in a typical use? For that matter, since it's new, do we have any idea what typical use will be? I'm thinking of some small number of users having maybe three or four parameters so maybe that biases my approach.  —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.